gobblin-data-management/src/main/java/gobblin/data/management/copy/ConcurrentBoundedWorkUnitList.java

```diff
@@ -0,0 +1,153 @@
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package gobblin.data.management.copy;
+
+import lombok.Getter;
+
+import java.util.Comparator;
+import java.util.List;
+import java.util.Set;
+import java.util.TreeMap;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Sets;
+
+import gobblin.data.management.partition.Partition;
+import gobblin.source.workunit.WorkUnit;
+
+
+/**
+ * A {@link WorkUnit} container that is bounded, supports concurrent all-or-nothing addAll, and supports priority of
+ * file sets, ie. attempting to add a file set with higher priority will automatically evict
+ * lower priority {@link Partition}s if necessary.
+ *
+ * <p>
+ *   File sets in {@link CopySource} are handled as {@link Partition}, so this class uses a {@link Partition} comparator
+ *   for priority. If fileSetA < fileSetB, then fileSetA has higher priority than fileSetB
+ *   (similar to {@link java.util.PriorityQueue}).
+ * </p>
+ */
+class ConcurrentBoundedWorkUnitList {
+
+  private final TreeMap<Partition<CopyableFile>, List<WorkUnit>> workUnitsMap;
+  @Getter
+  private final Comparator<Partition<CopyableFile>> comparator;
+  private final int maxSize;
+  private int currentSize;
+  /** Set to true the first time a file set is rejected (i.e. doesn't fit in the container) */
+  private boolean rejectedFileSet;
+
+  private class AugmentedComparator implements Comparator<Partition<CopyableFile>> {
+    private final Comparator<Partition<CopyableFile>> userProvidedComparator;
+
+    public AugmentedComparator(Comparator<Partition<CopyableFile>> userProvidedComparator) {
+      this.userProvidedComparator = userProvidedComparator;
+    }
+
+    @Override public int compare(Partition<CopyableFile> p1, Partition<CopyableFile> p2) {
+      int userProvidedCompare = this.userProvidedComparator.compare(p1, p2);
+      if (userProvidedCompare == 0) {
+        int datasetCompare = p1.getDataset().datasetRoot().compareTo(p2.getDataset().datasetRoot());
+        if (datasetCompare == 0) {
+          return p1.getName().compareTo(p2.getName());
+        }
+        return datasetCompare;
+      }
+      return userProvidedCompare;
+    }
+  }
+
+  /**
+   * Creates a new {@link ConcurrentBoundedWorkUnitList}.
+   * @param maxSize Maximum number of {@link WorkUnit}s to contain.
+   * @param comparator {@link Comparator} for {@link Partition}s to use for {@link Partition} priority.
+   */
+  public ConcurrentBoundedWorkUnitList(int maxSize, final Comparator<Partition<CopyableFile>> comparator) {
+    this.currentSize = 0;
+    this.maxSize = maxSize;
+    this.comparator = comparator;
+    this.workUnitsMap = new TreeMap<>(new AugmentedComparator(comparator));
+    this.rejectedFileSet = false;
+  }
+
+  /**
+   * Add a file set to the container.
+   * @param fileSet File set, expressed as a {@link Partition} of {@link CopyableFile}s.
```

Comment:
The reason is a previous discussion on the definition of "partition", and "file set" was a more agreeable term. I don't want to change the name file set, but we can change the name of the class `Partition` to `FileSet`.